feat: new `reference upload` command

emilyskuo opened the pull request:
| 🚥 Resolves RM-12965 |
| :------------------- |

## 🧰 Changes

Creating a new `reference upload` command for users to upload MD pages to the reference section. This is entirely based on the `docs upload` command, with most of it being exactly the same, minus the route and some typing.

I had looked into making this an alias of `docs upload`, but oclif doesn't tell you how the command was invoked, and we'd have to parse `argv`s to figure out if docs or reference, which could potentially work but would also be pretty brittle. Otherwise, we could create a general upload command for markdown pages and add a type arg, which would be changing up behavior even more.

## 🧬 QA & Testing

@kanadgupta consolidated our markdown upload command tests in https://github.com/readmeio/rdme/pull/1262 🎉

emilyskuo committed:
feat: add new `references upload` command for uploading MD to the reference section

emilyskuo committed:
fix: ajv error about a missing type

emilyskuo committed:
test: `references upload` command

emilyskuo committed:
test: update snapshots

emilyskuo commented on __tests__/__fixtures__/docs/existing-docs/subdir/another-doc.md:
```diff
@@ -1,7 +1,7 @@
 title: This is another document title
 category:
-  uri: /branches/stable/categories/guides/category-slug
+  uri: category-slug
```
Updating so we can use it with `references upload` as well

kanadgupta replied:
update: worked on this in https://github.com/readmeio/rdme/pull/1262!

emilyskuo commented on src/lib/types/openapiDoc.ts:
```diff
@@ -6525,7 +6525,7 @@ const document = {
                       },
                       path: { type: 'string', description: 'The endpoint path.' },
-                      schema: { nullable: true },
+                      schema: { type: 'string', nullable: true },
```
The lack of type here was breaking frontmatter validation with the error "nullable cannot be used without type" from ajv

erunion replied:
😬 I actually had to fix something similar for this over as part of my backgrounding work in https://github.com/readmeio/readme/pull/14881/commits/b09cb92c36831a1ca5ccc18fa035f8f663fc5319.

Since `type: string` is incorrect for this data shape, as it's a nullable object. Does `type: object` pass AJV here instead? Or if that doesn't work what about `additionalProperties: {}`?

```suggestion
                      schema: { type: 'object', properties: {}, nullable: true },
```

```suggestion
                      schema: { additionalProperties: {}, nullable: true },
```

Really wish I had just returned the schema as a string...

emilyskuo replied:
`schema: { type: 'object', properties: {}, nullable: true },` works!

erunion replied:
awesome, i'll see if i can update our OAS generation to make that change there so we don't have to worry about preserving it just here.

kanadgupta replied:
ope, just seeing this. yeah if we could defer to the auto-generated OAS as our source of truth, i'd prefer that approach. ([more info in this comment](https://github.com/readmeio/rdme/pull/1261#discussion_r2151158098))

emilyskuo committed:
chore: pr review comments

kanadgupta reviewed (changes requested):
this all looks great! the two blocking pieces of feedback should be fairly trivial: 
1. separating out the type fixes in the `src/lib/types/openapiDoc.ts` into its own PR
2. renaming the command from `rdme references upload` to `rdme reference upload`. could you also rename the two parent folders from "references" to "reference"?

kanadgupta commented on __tests__/commands/references/upload.test.ts:
update: figured this would help us move a little faster on support for custom pages, so i cranked this out in https://github.com/readmeio/rdme/pull/1262!

kanadgupta commented on __tests__/commands/references/upload.test.ts:
```diff
@@ -0,0 +1,822 @@
+const key = 'rdme_123';
+const authorization = `Bearer ${key}`;
+
+describe('rdme references upload', () => {
```
never mind, just merge this in instead: https://github.com/readmeio/rdme/pull/1262

emilyskuo committed:
chore: rename `references upload` to `reference upload`

emilyskuo commented on src/lib/types/openapiDoc.ts:
```diff
@@ -6525,7 +6525,7 @@ const document = {
                       },
                       path: { type: 'string', description: 'The endpoint path.' },
-                      schema: { nullable: true },
+                      schema: { type: 'object', properties: {}, nullable: true },
```
Gonna leave this as is as discussed during our sync

kanadgupta committed:
test: consolidate test suite for all markdown syncing commands (#1262)

## 🧰 Changes

this piggybacks off of @emilyskuo's work in
https://github.com/readmeio/rdme/pull/1261 and implements a suggestion i
had in [this
comment](https://github.com/readmeio/rdme/pull/1261/files#r2151162521)
(i.e., consolidating the tests for guides and reference uploads into a
single test suite wrapped in a `describe.each`). for now it consolidates
the tests for `rdme docs upload` and `rdme reference upload`, but i
figured it'll alleviate some of the burden of shipping support for
custom pages, etc. going forward.

as part of this, i did the following:

- [x] wrapped everything in a `describe.each`
- [x] added a bunch of template strings for the routes
- [x] duplicated a handful of fixtures
- [x] incorporated this feedback:
https://github.com/readmeio/rdme/pull/1261/files#r2151151087

## 🧬 QA & Testing

i went through all of the snapshots to make sure it all checks out, but
please feel free to look through everything yourself!

kanadgupta commented on src/commands/reference/upload.ts:
```diff
@@ -0,0 +1,89 @@
+    ...branchFlag(),
+    'confirm-autofixes': confirmAutofixesFlag,
+    'dry-run': Flags.boolean({
+      description: 'Runs the command without creating nor updating any References in ReadMe. Useful for debugging.',
```
```suggestion
      description: 'Runs the command without creating nor updating any reference pages in ReadMe. Useful for debugging.',
```

kanadgupta reviewed (approved):
one tiny copy edit below, otherwise LGTM! great work here ❤️‍🔥

emilyskuo committed:
Update src/commands/reference/upload.ts

Co-authored-by: Kanad Gupta <git@kanad.dev>

emilyskuo committed:
chore: lint

kanadgupta commented:
:tada: This PR is included in version 10.3.0-next.3 :tada:

The release is available on:
- [npm package (@next dist-tag)](https://www.npmjs.com/package/rdme/v/10.3.0-next.3)
- `v10.3.0-next.3`
- [GitHub release](https://github.com/readmeio/rdme/releases/tag/v10.3.0-next.3)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

kanadgupta commented:
:tada: This PR is included in version 10.3.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/rdme/v/10.3.0)
- `v10.3.0`
- [GitHub release](https://github.com/readmeio/rdme/releases/tag/v10.3.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
